### PR TITLE
 fix rerun before/after hooks on navigation [remerge]

### LINKED ISSUE #1987: Code in hooks and test are rerun when changing to a different superdomain
### Current behavior:

When running my spec file, I noticed that some code in the `before` hook (not `beforeEach`) is being run before each of my three test cases, causing my checks to fail. Want to remark that I am using different superdomains for every test, I don't see this when I run all the tests with the same superdomain.

### Desired behavior:

Code in the `before` hook only runs once, before all the tests, no matter if using different superdomains.

### Steps to reproduce:

I explain my setup. My spec file consists of:
```
- before()
- after()
- it() //superdomain 1
- it() //superdomain 2 using an iframe
- it() //superdomain 1
```
As I am using different superdomains, the only way that I've read to do it is visiting the different domains in different tests (https://docs.cypress.io/guides/guides/web-security.html#One-Superdomain-per-Test)

I've also set `"chromeWebSecurity": false`

Within the before hook I run several requests to some endpoints as a way to get some data that I will use through the tests; Get it once, and use it across the three it().

I've also set a `console.log('This only happens in the beforeHook:  ' + var)` to see what's the value of my variable, and this is what I see in the run:

- Running the first it()

 
- Second it()

- Third it()

After every `it()` the console gets cleared, which is fine, but I don't expect to see the log again during the second and the third tests, as this is supposed to run only once. Notice how the variable gets an increment for each loop, as I would expect had I done it in a `beforeEach` hook.

In the left side of the window every seems fine as there is no command logging for the `before` hook when running the second and third tests, but the `console.log` reveals that.

### Versions

- Cypress version: 3.0.1
- Browser: Chrome 66
- OS: Ubuntu Linux 16.04 LTS

### LINKED ISSUE #2296: Exception in root level after() doesn't fail test run in interactive mode
### Current behavior:

If there's an `after()` outside of any `describe`, `it` or `context` ("rool level"), which throws an exception, that doesn't fail the test run. This is different from how it works in `mocha`, where you get a `"after all" hook` failure.

Also, if it's a synchronous failure, the exception doesn't show up anywhere.
If it's async, it shows up like this:

screen shot 2018-08-07 at 5 13 10 pm

### Desired behavior:

The whole test run should fail (sometimes you would want to place an assertion there that is possible only after all tests have run).

### Steps to reproduce:

Add this to your tests:
```js
after(() => {
  cy.then(() => {
    throw new Error('bla bla');
  });
});
```

### Versions

Cypress: 3.0.3
OS: Mac OS X 10.12.6 (Sierra)
Browser: both Electron (59) and Chrome (68)

### PR DESCRIPTION
**Pull request by kuceb**
prev PR accidentally merged #7035 

- fix #1987
- fix #2296 failure in root after hook doesn't show failure in browser reporter
- address #2777 (fixes half of the problem - where hooks are rerun on navigation)

### User facing changelog
- Bug fix: on domain navigation, `before`/`after` hooks defined in completed suites no longer erroneously rerun
- Bug fix: command logs from root `after` hooks are now displayed on the correct test

#### Before: Errors in root `after` hooks wouldn't fail test

Screen Shot 2020-04-28 at 12 52 59 PM

#### After: Errors in root `after` hooks fail test

Screen Shot 2020-04-28 at 12 54 41 PM

### Tasks
- [x] cleanup
- [x] wait for merge #7486 

### PR Tasks
- [x] rename duplicate issue number
- [x] add testcase for navigation inside suite
- [x] this introduced another bug due to other broken behavior in cypress related to global hooks, so this currently skips global after hooks after navigation - add small exclusion for this edge case
- [x] remove an old conditional around predicting when the last afterhook has fired, which fixes the hook firing order bug

- [x] Have tests been added/updated?
- [x] Has the original issue been tagged with a release in ZenHub?

### CODE REVIEW

File: packages/server/test/support/fixtures/projects/hooks-after-rerun/cypress/support/index.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+// ***********************************************************
+// This example support/index.js is processed and
+// loaded automatically before your test files.
+//
```
Comment (jennifer-shehane) [inline packages/server/test/support/fixtures/projects/hooks-after-rerun/cypress/support/index.js:4]: you should set `supportFile: false` in the cypress config and delete these files. https://github.com/cypress-io/cypress/blob/2-fix-rerun-hooks-on-navigation/packages/server/test/support/fixtures/projects/hooks-after-rerun/cypress.json

**Review by jennifer-shehane — COMMENTED** [review]
I manually confirmed this fixes the issues mentioned.

**Comment by jennifer-shehane** [comment]
@chrisbreiding was asking https://github.com/cypress-io/cypress/pull/7035#issuecomment-622524925

>Does this also fix #3904?

**Comment by jennifer-shehane** [comment]
Was trying to fix if this PR fixes an issue, but running into this error and the tests don't run.

Screen Shot 2020-05-26 at 3 05 47 PM

**Review by brian-mann — DISMISSED** [review]

**Review by brian-mann — APPROVED** [review]

**Comment by vineetbrikl** [comment]
Hi,
Is the issue with before/after hook rerun on superdomain navigation resolved? It was mentioned in #1987 that it has been released with version 4.8.0. I tried in both 6.6.0 and 4.80 I am still able to reproduce. There seems to be multiple PRs regarding this issue but there's not clear information about it was actually fixed or not. Could you please confirm when this is going to get fixed?

**Pull request merged**